lib/src/connector/connector_base.cc

```diff
@@ -0,0 +1,289 @@
+    }
+}
+
+void ConnectorBase::stopMonitoring()
+{
+    checkConnectionInitialization();
+
+    if (is_monitoring_) {
+        stopMonitorTaskAndWait();
+    } else if (monitor_exception_) {
+        LOG_DEBUG("The Monitoring Thread previously caught an exception; "
+                  "re-throwing it");
+        boost::rethrow_exception(monitor_exception_);
+    } else {
+        LOG_WARNING("The Monitoring Thread is not running");
+    }
+}
+
+void ConnectorBase::monitorConnection(const uint32_t max_connect_attempts,
+                                  const uint32_t connection_check_interval_s)
+{
+    checkConnectionInitialization();
+    checkPingTimings(connection_check_interval_s*1000, client_metadata_.pong_timeout_ms);
+
+    if (!is_monitoring_) {
+        is_monitoring_ = true;
+        startMonitorTask(max_connect_attempts, connection_check_interval_s);
+
+        // If startMonitorTask exits because of an exception, rethrow it now.
+        // Avoid a race condition if the thread is stopped asynchronously by
+        // checking must_stop_monitoring_.
+        if (!must_stop_monitoring_ && monitor_exception_) {
+            boost::rethrow_exception(monitor_exception_);
+        }
+    } else {
+        LOG_WARNING("The Monitoring Thread is already running");
+    }
+}
+
+// Utility functions
+
+void ConnectorBase::checkConnectionInitialization()
+{
+    if (connection_ptr_ == nullptr) {
+        throw connection_not_init_error {
+            lth_loc::translate("connection not initialized") };
+    }
+}
+
+//
+// Monitoring Task
+//
+
+void ConnectorBase::startMonitorTask(const uint32_t max_connect_attempts,
+                                 const uint32_t connection_check_interval_s)
+{
+    assert(connection_ptr_ != nullptr);
+    // Reset the exception, in case one was previously triggered and handled.
+    monitor_exception_ = {};
+    LOG_INFO("Starting the monitor task");
+    Util::chrono::system_clock::time_point now {};
+    Util::unique_lock<Util::mutex> the_lock { monitor_mutex_ };
+
+    while (!must_stop_monitoring_) {
+        now = Util::chrono::system_clock::now();
+
+        monitor_cond_var_.wait_until(
+            the_lock,
+            now + Util::chrono::seconds(connection_check_interval_s));
+
+        if (must_stop_monitoring_)
+            break;
+
+        try {
+            if (!isConnected()) {
+                LOG_WARNING("WebSocket connection to PCP broker lost; retrying");
+                connect(max_connect_attempts);
+            } else {
+                LOG_DEBUG("Sending heartbeat ping");
+                connection_ptr_->ping();
+            }
+        } catch (const connection_config_error& e) {
+            // Connection::connect(), ping() or WebSocket TLS init
+            // error - retry
+            LOG_WARNING("The connection monitor task will continue after a "
+                        "WebSocket failure ({1})", e.what());
+        } catch (const connection_association_error& e) {
```

Comment:
Is it worth to avoid having references to Session Association here?

You could move this logic to the child classes by making the monitor task a pure virtual function.

Or you could distinguish between connection errors that do / don't require reconnection attempts and keep the implemenation in the base class. By the way, following the v2 cleanup, the whole set of error types could benefit from some simplification; too many of them...

Either way, if worth the time, this can be solved in a future PR.